pkg/logstructured/logstructured.go

```diff
@@ -291,16 +291,16 @@ func (l *LogStructured) ttl(ctx context.Context) {
 				continue
 			}
 
-			expiredAt := time.Now().Add(time.Duration(event.KV.Lease) * time.Second)
 			eventKV := loadTTLEventKV(rwMutex, ttlEventKVMap, event.KV.Key)
 			if eventKV == nil {
 				logrus.Tracef("Add ttl event key %v, modRev %v", event.KV.Key, event.KV.ModRevision)
-				storeTTLEventKV(rwMutex, ttlEventKVMap, event.KV.Key, event.KV.ModRevision, expiredAt)
-				queue.AddAfter(event.KV.Key, time.Duration(event.KV.Lease)*time.Second)
+				expires := storeTTLEventKV(rwMutex, ttlEventKVMap, event.KV)
```

Comment:
> just passing it in here during the store won't fix anything; you'd need to get the lock before loading the the entry, and hold it until after the checks are done.
> 
> I'm not even convinced that it's necessary, I just wanted to think about whether or not this creates the possibility of a race condition.
> 
> If it does matter, and we need to hold a lock while checking the existing entry, then we probably need to make a type that holds both the map and the lock, and have a function on that type to handle locking and comparison - rather than exposing it here.

I don't think there will be any problems here. First of all, the event processing logic is handled by a single process, Secondly, even if the load retrieves the previous deleted element, the event. ModRevision will be compared during the update. If it is a new event, the new event will be queued.

Reply:
I agree, I think we're probably fine without a lock.